Turin TCB support

jraman567 opened the pull request:
Turin and later families have FMC SPL in their TCBs. Additionally, the HwID in the TCB is only 8 bytes long for Turin and later, compared to 64 bytes or Genoa and earlier.

Please see section 3.1 in the spec below:
https://www.amd.com/content/dam/amd/en/documents/epyc-technical-docs/specifications/57230.pdf

jraman567 committed:
Turin TCB support

Turin and later families have FMC SPL in their TCBs. Additionally, the HwID
in the TCB is only 8 bytes long for Turin and later, compared to 64 bytes
or Genoa and earlier.

Please see section 3.1 in the spec below:
https://www.amd.com/content/dam/amd/en/documents/epyc-technical-docs/specifications/57230.pdf

Signed-off-by: Jagannathan Raman <jraman567@gmail.com>

jraman567 commented on kds/kds.go:
```diff
@@ -114,17 +127,23 @@ var (
-// TCBVersion is a 64-bit bitfield of different security patch levels of AMD firmware and microcode.
-type TCBVersion uint64
+// TCBVersion represents the SEV-SNP TCB_VERSION
+type TCBVersion struct {
+	// Version identifies the version of this structure as defined in V[C,L]EK x509 certificate extensions
+	Version uint8
+	// TCB is a 64-bit bitfield of different security patch levels of AMD firmware and microcode.
+	TCB uint64
+}
```
kds.DecomposeTCBVersion() defaults to structVersion0 to ensure it doesn't break the API

deeglaze replied:
kds.TCBVersion(0) is a type error with its change to a struct. Perhaps we do need to go back to interfaces, but ones that aren't specific to the SPL naming that exists for Turin.

deeglaze replied:
I don't think this is significant enough to warrant a major version change. You can create an issue for cleaning it up and label it for major-release. There are a fair number of exported functions that were not meant for public consumption, but rather internal consumption across packages, so we should collect up that wishlist before going to v1.

jraman567 replied:
OK, thanks for confirming @deeglaze . If I understand correctly, we are going to proceed with this PR and shelve the [other PR](https://github.com/google/go-sev-guest/pull/169).

I'll file an issue to clean this up.

jraman567 replied:
@deeglaze Thanks for reviewing so far! I've addressed the backwards compatibility issue you brought up.

jraman567 replied:
Marking this conversation as resolved as we've addressed the backwards compatibility issue

larrydewey commented on kds/kds.go:
```diff
@@ -382,7 +449,22 @@ func kdsOidMapToExtensions(exts map[kdsOID]*pkix.Extension) (*Extensions, error)
+
+	switch result.StructVersion {
+	case TCBStructVersion0:
+		fmcspl = 0
```
We cannot rely on Go's default values for FMC in the unsupported case as 0 is a valid version number. Perhaps we can introduce a `nil` here?

jraman567 replied:
This one just ensures that the `fmcspl` variable is initialized to 0. It is only used if TCB version structure is structVersion1. Otherwise, it's ignored.

jraman567 replied:
@larrydewey It was just initializing the unused variables to 0. But I see how it can be confusing.

I've addressed it and marking this conversation as resolved. If anything changes, please let me know.

jraman567 committed:
address feedback from review

- fix lint errors reported by Ubuntu CI
- make error messages in tests reader friendly
- address Larry's feedback
- address backwards compatibility issues raised in the review

Signed-off-by: Jagannathan Raman <jraman567@gmail.com>

larrydewey commented on kds/kds.go:
```diff
@@ -208,11 +232,14 @@ type TCBParts struct {
 	// UcodeSpl is the microcode security patch level.
 	UcodeSpl uint8
+	// FmcSpl is the FMC security patch level.
+	FmcSpl uint8
```
I see two potential approaches to solving the zero value issue here:

## Option 1:

Utilize a pointer instead of a value:

```suggestion
	FmcSpl *uint8
```

In this case we can assign `nil` to the value when it isn't present.

---

## Option 2:

Define a custom `Option` type similar to what is available in Rust.

```go
type Option[T any] struct {
	value    T
	hasValue bool
}

func Some[T any](value T) Option[T] {
	return Option[T]{value: value, hasValue: true}
}

func None[T any]() Option[T] {
	return Option[T]{}
}

func (o *Option[T]) IsSome() bool {
	return o.hasValue
}

func (o *Option[T]) Unwrap() T {
	if !o.hasValue {
		panic("Unwrap on None")
	}
	return o.value
}
```

```suggestion
	FmcSpl Option[uint8]
```

My guess is the easiest implementation would be to just use option 1, but I will leave that up to you.

jraman567 replied:
Thanks for the suggestions.

larrydewey commented on kds/kds.go:
```diff
@@ -227,32 +254,63 @@ func ComposeTCBParts(parts TCBParts) (TCBVersion, error) {
+			(uint64(parts.TeeSpl) << 16) |
+			(uint64(parts.BlSpl) << 8) |
+			(uint64(parts.FmcSpl) << 0)
+	default:
+		return tcbVersion, fmt.Errorf("unsupported TCB struct version: %v", parts.Version)
+	}
```
I don't anticipate the need for an error here. It should be safe to default to one structure if it isn't explicitly the other version. @deeglaze thoughts on this?

larrydewey commented on kds/kds.go:
```diff
@@ -332,17 +391,28 @@ func asn1OctetString(ext *pkix.Extension, field string, size int) ([]byte, error
+		hwIdLen = int(TCBHwIdLenVersion1)
+	default:
+		return nil, fmt.Errorf("unsupported TCB structVersion %d", result.StructVersion)
+	}
```
Same thought here. Can we not just default to one version?

larrydewey commented on kds/kds.go:
```diff
@@ -426,7 +509,7 @@ func VcekCertificateExtensions(cert *x509.Certificate) (*Extensions, error) {
 		return nil, fmt.Errorf("unexpected CSP_ID in VCEK certificate: %s", exts.CspID)
 	}
-	if len(exts.HWID) != abi.ChipIDSize {
+	if len(exts.HWID) != int(TCBHwIdLenVersion0) && len(exts.HWID) != int(TCBHwIdLenVersion1) {
```
You can save a few CPU instructions here.
```suggestion
        var extsLen int = len(exts.HWID)
	if extsLen != int(TCBHwIdLenVersion0) && extsLen != int(TCBHwIdLenVersion1) {
```

larrydewey commented on kds/kds.go:
```diff
@@ -514,27 +597,57 @@ func ProductCertChainURL(s abi.ReportSigner, productLine string) string {
+			parts.UcodeSpl,
+			parts.FmcSpl,
+		)
+	case TCBStructVersion0:
+		fallthrough
```
Again, I am not sure this is needed at all.

larrydewey commented on kds/kds.go:
```diff
@@ -514,27 +597,57 @@ func ProductCertChainURL(s abi.ReportSigner, productLine string) string {
+			parts.UcodeSpl,
+			parts.FmcSpl,
+		)
+	case TCBStructVersion0:
+		fallthrough
```
Likely not needed

larrydewey commented on kds/kds.go:
```diff
@@ -671,7 +786,20 @@ func parseTCBURL(u *url.URL) (uint64, error) {
+// ProductLineToTCBVersion maps the product line to the TCB version they use
+// in V{C,L}EK x509 extensions
+func ProductLineToTCBVersion(productLine string) (uint8, error) {
+	switch productLine {
```
Not sure where this is being called, but we definitely want to normalize this.
```suggestion
	switch strings.Title(productLine) {
```

jraman567 replied:
I'm using `cases.Title` instead as `strings.Title` appears to be deprecated.

larrydewey commented on tools/lib/report/report.go:
```diff
@@ -121,17 +121,31 @@ func asBin(report *spb.Attestation) ([]byte, error) {
+	case kds.TCBStructVersion1:
+		return fmt.Sprintf("0x%x:{ucode: %d, snp: %d, tee: %d, bl: %d, fmc: %d}",
+			tcb, parts.UcodeSpl, parts.SnpSpl, parts.TeeSpl, parts.BlSpl, parts.FmcSpl)
+	case kds.TCBStructVersion0:
+		fallthrough
```
I don't think we need to have this, as default covers it.

larrydewey commented on tools/lib/report/report.go:
```diff
@@ -121,17 +121,31 @@ func asBin(report *spb.Attestation) ([]byte, error) {
-		tcbBreakdown(report.Report.GetCurrentTcb()),
-		tcbBreakdown(report.Report.GetCommittedTcb()),
-		tcbBreakdown(report.Report.GetLaunchTcb()))), nil
+		tcbBreakdown(tcbVersion, report.Report.GetCurrentTcb()),
+		tcbBreakdown(tcbVersion, report.Report.GetCommittedTcb()),
+		tcbBreakdown(tcbVersion, report.Report.GetLaunchTcb()))), nil
```
Instead of passing this multiple times, would it make sense to call `ProductLineToTCBVersion` inside of tcbBreakdown?

jraman567 replied:
We anyway have to call tcbBreakdown for Current, Committed and Launch TCBs.

I don't have a preference, we can do it either way. If we do  `ProductLineToTCBVersion` inside `tcbBreakdown`, then we have to pass the product name to it instead of tcbVersion.

jraman567 replied:
@larrydewey I'll mark this conversation as resolved if you don't have objection to the existing approach. Please re-open it if you want to discuss this further. Thank you!

larrydewey commented on validate/validate.go:
```diff
@@ -228,6 +230,16 @@ func PolicyToOptions(policy *cpb.Policy) (*Options, error) {
+	switch policy.Product.Name {
+	case spb.SevProduct_SEV_PRODUCT_TURIN:
+		tcbVersion = kds.TCBStructVersion1
+	case spb.SevProduct_SEV_PRODUCT_MILAN, spb.SevProduct_SEV_PRODUCT_GENOA:
+		tcbVersion = kds.TCBStructVersion0
```
Has this been tested with Siena and Bergamo?

jraman567 replied:
Thanks for the pointers! I haven't explicitly tested with Sienna and Bergamo, but it looks like we should as they are also part of the Zen4 architecture like Genoa.

larrydewey commented on kds/kds.go:
```diff
@@ -114,17 +127,23 @@ var (
-type TCBVersion uint64
+// TCBVersion represents the SEV-SNP TCB_VERSION
+type TCBVersion struct {
+	// Version identifies the version of this structure as defined in V[C,L]EK x509 certificate extensions
+	Version uint8
```
Where the version is purely for internal parsing and handling, does this need to be public?

larrydewey replied:
Fair enough. There are definitely ways of obfuscating this, but it is probably outside of the scope of this PR.

larrydewey reviewed (changes requested):
Some comments and questions

jraman567 commented on kds/kds.go:
```diff
@@ -382,7 +449,22 @@ func kdsOidMapToExtensions(exts map[kdsOID]*pkix.Extension) (*Extensions, error)
+		}
+	case TCBStructVersion1:
+		spl4 = 0
+		if err := asn1U8(exts[kdsUcodeSpl], "FmcSpl", &fmcspl); err != nil {
```
That's a good catch, thanks!

deeglaze commented on kds/kds.go:
```diff
@@ -192,6 +214,8 @@ func kdsOidMap(cert *x509.Certificate) (map[kdsOID]*pkix.Extension, error) {
 // an AMD secure processor firmware TCB version.
 type TCBParts struct {
+	// Version is the version of the TCB structure
+	Version uint8
```
It's hard to put the cat back inside the bag. I wouldn't say this PR would get no attention if it's not merged before Juneteenth. Is there no way your service can work from a fork until this PR is clean?

jraman567 replied:
OK, yeah, makes sense. It's best to get it right the first time.

I'll update the PR with changes to make version private.

deeglaze commented on kds/kds.go:
```diff
@@ -227,33 +323,104 @@ func ComposeTCBParts(parts TCBParts) (TCBVersion, error) {
+	return TCBVersion(tcb.TCB), nil
+}
+
+// ComposeTCBPartsToTCBVersionStruct returns an SEV-SNP TCB_VERSION from OID mapping
+// values as a TCBVersionStruct. The spl4-spl7 fields are reserved, but the
+// KDS specification designates them as 4 byte-sized fields.
+func ComposeTCBPartsToTCBVersionStruct(parts TCBParts) (TCBVersionStruct, error) {
+	return parts.ToTCBVersionStruct()
```
Why is this function necessary if you're exporting ToTCBVersionStruct?

jraman567 replied:
Sorry about that, I meant to remove it.

deeglaze commented on kds/kds.go:
```diff
@@ -59,12 +61,24 @@ var (
 	// origin TLS key's certificate's subject key's CommonName.
 	OidCspID = asn1.ObjectIdentifier([]int{1, 3, 6, 1, 4, 1, 3704, 1, 5})
 
+	// TCBStructVersion0 is the version of the TCB structure for Milan & Genoa
+	TCBStructVersion0 = uint8(0)
+	// TCBStructVersion1 is the version of the TCB structure for Turin
+	TCBStructVersion1 = uint8(1)
```
Struct version constants shouldn't need to be exported given the version fields aren't exported.

deeglaze commented on kds/kds.go:
```diff
@@ -227,33 +323,104 @@ func ComposeTCBParts(parts TCBParts) (TCBVersion, error) {
+	case TCBVersion:
+		tcb.FromTCBVersion(v)
+	case *TCBVersion:
+		if v == nil {
+			return TCBParts{}, fmt.Errorf("TCBVersion is nil")
+		}
```
can't the nil comparison be lifted out of the switch?

jraman567 replied:
ACK

jraman567 commented:
@deeglaze I've addressed your recent comments

deeglaze commented on kds/kds.go:
```diff
@@ -115,19 +133,94 @@ var (
+	switch c.String(productLine) {
+	case "Turin":
+		return &TCBVersionStruct{version: tcbStructVersion1, TCB: tcb}, nil
+	default:
```
This should fail closed. We don't want "Venice" to accidentally end up misinterpreted here.

jraman567 replied:
@deeglaze When I do that, some of the validate tests are failing. The root cause is that the Reports generated by the QuoteProvider doesn't have the FMS set. Do you know if we could configure the QuoteProvider to fill the FMS in the report?

jraman567 replied:
The issue is the report generated in the following code doesn't have FMS in the report:
```go
		{
			name: "just reportData",
			attestation: func() *spb.Attestation {
				q, err := sg.GetQuoteProto(qp0, nonce0s1)
				if err != nil {
					t.Fatal(err)
				}
				report := q.Report
				//report.Cpuid1EaxFms = 0xa00f11
				return &spb.Attestation{
					Report: report,
					CertificateChain: &spb.CertificateChain{
						AskCert:  sign0.Ask.Raw,
						ArkCert:  sign0.Ark.Raw,
						VcekCert: sign0.Vcek.Raw,
					},
				}
			}(),
			opts: &Options{ReportData: nonce0s1[:], GuestPolicy: abi.SnpPolicy{Debug: true}},
		},
```

deeglaze replied:
Yes, we can use test data that depends on version 3 and populates the FMS.

jraman567 replied:
I could add file and track an issue to use v3 reports in tools/check and validate_test, if you're OK with defaulting to structVersion0 here.

deeglaze reviewed (approved):
LGTM with final comments.

onlyshawn committed:
Merge branch 'main' into turin-tcb-v2

muhmud commented:
Any updates on when this would be reviewed & merged?
